phoenix-core/src/main/java/org/apache/phoenix/mapreduce/VerifyReplicationTool.java

```diff
@@ -0,0 +1,477 @@
+                        keyCompare = 1;
+                    } else {
+                        keyCompare = Bytes.compareTo(value.getSourceKey().get(), targetKey.get());
+                    }
+                    if(keyCompare == 0) {
+                        Map<String, Object> sourceValues = value.getSourceResults();
+                        Map<String, Object> targetValues = null;
+                        PhoenixRecordWritable targetResults = new PhoenixRecordWritable();
+                        targetResults.readFields(targetResultSet);
+                        if (targetResults.getResultMap() != null) {
+                            targetValues = targetResults.getResultMap();
+                        }
+                        boolean valuesMatch = true;
+                        if (sourceValues == null) {
+                            if (targetValues != null) {
+                                valuesMatch = false;
+                            }
+                        } else if (!sourceValues.equals(targetValues)) {
+                            valuesMatch = false;
+                        }
+                        if (!valuesMatch) {
+                            logFailRowAndIncrementCounter(context, Counter.CONTENT_DIFFERENT_ROWS, value.getSourceKey());
+                        } else {
+                            context.getCounter(Counter.GOODROWS).increment(1);
+                        }
+                        targetHasNext = targetResultSet.next();
+                        break;
+                    } else if (keyCompare < 0) {
+                        // row only exists in source table
+                        logFailRowAndIncrementCounter(context, Counter.ONLY_IN_SOURCE_TABLE_ROWS, value.getSourceKey());
+                        break;
+                    } else {
+                        // row only exists in target table
+                        logFailRowAndIncrementCounter(context, Counter.ONLY_IN_TARGET_TABLE_ROWS,
+                                targetKey);
+                        targetHasNext = targetResultSet.next();
+                    }
+                }
+            } catch (SQLException e) {
+                throw new IOException(e.getMessage());
+            }
+        }
+
+        @Override
+        protected void cleanup(Context context) throws IOException, InterruptedException {
+            super.cleanup(context);
+            try {
+                if (!sourceHasData && targetResultSet == null) {
+                    getTargetResultSet(context.getConfiguration(), null, null);
+                }
+                if(targetResultSet != null) {
+                    while (targetHasNext) {
+                        ImmutableBytesWritable targetKey = null;
+
+                        if (targetResultSet.getCurrentRow() != null) {
+                            targetKey = new ImmutableBytesWritable();
+                            targetResultSet.getCurrentRow().getKey(targetKey);
```

Comment:
We are allocating new byte[] for every row. Can this potentially cause OOME inside a container if the size of region is large? Could you reuse these buffer? @apurtell @aaraujo

Reply:
ImmutableByteWritable does not allocate a new byte[]. Also I think this is only called if there are more rows on the target than the source for this map task.

Reply:
@twdsilva That's correct.

Reply:
> ImmutableByteWritable does not allocate a new byte[]. Also I think this is only called if there are more rows on the target than the source for this map task.

Yes, I saw that. The line 221 will allocate a new byte[] for every row though. I assume that we might not have a huge deviation between rows, so it should be fine.